### PR TITLE
[SPARK-29975][SQL] introduce --CONFIG_DIM directive

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

allow the sql test files to specify different dimensions of config sets during testing. For example,
```
--CONFIG_DIM1 a=1
--CONFIG_DIM1 b=2,c=3

--CONFIG_DIM2 x=1
--CONFIG_DIM2 y=1,z=2
```

This example defines 2 config dimensions, and each dimension defines 2 config sets. We will run the queries 4 times:
1. a=1, x=1
2. a=1, y=1, z=2
3. b=2, c=3, x=1
4. b=2, c=3, y=1, z=2

### Why are the changes needed?

Currently `SQLQueryTestSuite` takes a long time. This is because we run each test at least 3 times, to check with different codegen modes. This is not necessary for most of the tests, e.g. DESC TABLE. We should only check these codegen modes for certain tests.

With the --CONFIG_DIM directive, we can do things like: test different join operator(broadcast or shuffle join) X different codegen modes.

After reducing testing time, we should be able to run thrifter server SQL tests with config settings.

### Does this PR introduce any user-facing change?

no

### How was this patch tested?

test only

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @maropu @wangyum @MaxGekk @dongjoon-hyun

File: sql/core/src/test/resources/sql-tests/inputs/join-empty-relation.sql [hunk 1]
```diff
@@ -1,8 +1,3 @@
--- List of configuration the test suite is run against:
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/join-empty-relation.sql]: This is to test the optimizer, don't need to run it with different join operators

File: sql/core/src/test/resources/sql-tests/inputs/natural-join.sql [hunk 2]
```diff
@@ -1,8 +1,3 @@
--- List of configuration the test suite is run against:
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/natural-join.sql]: This is to test the analyzer/optimizer. Natural join will be rewritten to other normal joins.

File: sql/core/src/test/resources/sql-tests/inputs/outer-join.sql [hunk 3]
```diff
@@ -29,16 +36,10 @@ CREATE OR REPLACE TEMPORARY VIEW t1 AS SELECT * FROM VALUES (97) as t1(int_col1)
 
 CREATE OR REPLACE TEMPORARY VIEW t2 AS SELECT * FROM VALUES (0) as t2(int_col1);
 
--- Set the cross join enabled flag for the LEFT JOIN test since there's no join condition.
--- Ultimately the join should be optimized away.
-set spark.sql.crossJoin.enabled = true;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/outer-join.sql]: this is true by default now.

File: sql/core/src/test/resources/sql-tests/inputs/udf/udf-join-empty-relation.sql [hunk 4]
```diff
@@ -1,8 +1,3 @@
--- List of configuration the test suite is run against:
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/udf/udf-join-empty-relation.sql]: we are testing UDFs, and the join operator doesn't matter.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerQueryTestSuite.scala [hunk 5]
```diff
@@ -104,9 +101,6 @@ class ThriftServerQueryTestSuite extends SQLQueryTestSuite {
     "subquery/in-subquery/simple-in.sql",
     "subquery/in-subquery/in-order-by.sql",
     "subquery/in-subquery/in-set-operations.sql",
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerQueryTestSuite.scala]: Hi, @cloud-fan .
This causes linter failure. Please remove the ending `,`.
```
$ dev/lint-scala
Scalastyle checks failed at following occurrences:
[error] /Users/dongjoon/PRS/PR-26612/sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerQueryTestSuite.scala: illegal start of simple expression: Token(RPAREN,),3495,))
[error] Total time: 20 s, completed Nov 20, 2019 11:24:55 AM
```

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 6]
```diff
@@ -271,36 +260,36 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession {
...
-    // `--SET` will be respected in this case.
-    if ((regenerateGoldenFiles && importedTestCaseName.isEmpty) || !isTestWithConfigSets) {
-      runQueries(queries, testCase, None)
+    val settingLines = comments.filter(_.startsWith("--SET ")).map(_.substring(6))
+    val settings = settingLines.flatMap(_.split(",").map { kv =>
+      val (conf, value) = kv.span(_ != '=')
+      conf.trim -> value.substring(1).trim
+    })
+
+    if (regenerateGoldenFiles) {
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: We can remove this line:
https://github.com/apache/spark/blob/3be4734f317b3b26da278ad088b31ec4b7750213/sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala#L114

**Comment by maropu** [comment]
retest this please

**Comment by cloud-fan** [comment]
After this PR:
- SQLQueryTestSuite: 25min
- ThriftServerQueryTestSuite: 18min

Before this PR: (use the result of https://github.com/apache/spark/pull/26214)
- SQLQueryTestSuite: 47min
- ThriftServerQueryTestSuite: 8min (but can't apply settings)

**Comment by wangyum** [comment]
Yes. Much faster than before.
SQLQueryTestSuite:
```
[info] Run completed in 15 minutes, 43 seconds.
[info] Total number of tests run: 202
```
ThriftServerQueryTestSuite:
```
[info] Run completed in 13 minutes, 13 seconds.
[info] Total number of tests run: 137
```

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql [hunk 7]
```diff
@@ -1,3 +1,8 @@
+-- Test aggregate operator with codegen on and off.
+--CONFIG_DIM1 spark.sql.codegen.wholeStage=true
+--CONFIG_DIM1 spark.sql.codegen.wholeStage=false,spark.sql.codegen.factoryMode=CODEGEN_ONLY
+--CONFIG_DIM1 spark.sql.codegen.wholeStage=false,spark.sql.codegen.factoryMode=NO_CODEGEN
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/group-by.sql]: (Sorry to be late) In the one dimension case, `CONFIG_DIM1` is the same with `SET`?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 8]
```diff
@@ -271,36 +259,36 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession {
...
-        val configLines = comments.filter(_.startsWith("--SET ")).map(_.substring(6))
-        val configs = configLines.map(_.split(",").map { confAndValue =>
-          val (conf, value) = confAndValue.span(_ != '=')
+      // A config dimension has multiple config sets, and a config set has multiple configs.
+      // - config dim:     Seq[Seq[(String, String)]]
+      //   - config set:   Seq[(String, String)]
+      //     - config:     (String, String))
+      // We need to do cartesian product for all the config dimensions, to get a list of
+      // config sets, and run the query once for each config set.
+      val configDimLines = comments.filter(_.startsWith("--CONFIG_DIM")).map(_.substring(12))
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: Better to update how-to use `CONFIG_DIM` in https://github.com/apache/spark/blob/6146dc4562739c1c947eb944897c2fe85d1016e0/sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala#L69

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: Yeah we should document this ...

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql (hunk 7 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/group-by.sql]: it's different. We still run this test 3 times as there are 3 config sets in this dimension. It's only the same with SET if there is only one dimension and one config set.

**Pull request closed without merging**